src/Data/Group/Permutation.hs

```diff
@@ -55,20 +81,47 @@ data Permutation a = Permutation
 -- instance Profunctor Permutation where
 --   dimap = :'(
 
-instance Semigroup a => Semigroup (Permutation a) where
-  a <> b = Permutation (to a <> to b) (from a <> from b)
+instance Semigroup (Permutation a) where
+  a <> b = Permutation (to a . to b) (from b . from a)
 
-instance Monoid a => Monoid (Permutation a) where
+instance Monoid (Permutation a) where
   mempty = Permutation id id
 
-instance Group a => Group (Permutation a) where
-  invert (Permutation t f) = Permutation (f . t) (t . f)
-
-instance Abelian a => Abelian (Permutation a)
-instance AdditiveGroup a => AdditiveGroup (Permutation a)
-instance AdditiveAbelianGroup a => AdditiveAbelianGroup (Permutation a)
-instance MultiplicativeGroup a => MultiplicativeGroup (Permutation a)
-
+instance Group (Permutation a) where
+  invert (Permutation t f) = Permutation f t
+
+instance (Enum a, Bounded a) => Eq (Permutation a) where
+  (==) = (==) `on` (functionRepr . to)
+
+instance (Enum a, Bounded a) => Ord (Permutation a) where
+  compare = compare `on` (functionRepr . to)
+
+instance (Enum a, Bounded a) => GroupOrder (Permutation a) where
+  order Permutation{to = f} = Finite (go 1 fullSet)
+    where
+      n = 1 + fromEnum (maxBound @a)
+      fullSet = ISet.fromDistinctAscList [0 .. n - 1]
+      
+      f' :: Int -> Int
+      f' = fromEnum . f . toEnum
+      
+      go :: Natural -> ISet.IntSet -> Natural
+      go !ord elements = case ISet.minView elements of
+        Nothing             -> ord
+        Just (k, elements') ->
+          let (period, elements'') = takeCycle k elements'
+          in go (lcm period ord) elements''
+      
+      takeCycle :: Int -> ISet.IntSet -> (Natural, ISet.IntSet)
+      takeCycle k = loop 1 (f' k)
+        where
+          loop !period j elements
+            | j `ISet.member` elements      = loop (succ period) (f' j) (ISet.delete j elements)
+            | {- j ∉ elements && -} j == k = (period, elements)
+            | otherwise                     = error $ "Non-bijective: witness=toEnum " ++ show j
```

Comment:
```suggestion
equalPermutation
  :: (Enum a, Bounded a) => Permutation a -> Permutation a -> Bool
equalPermutation = (==) `on` (functionRepr . to)

comparePermutation
  :: (Enum a, Bounded a) => Permutation a -> Permutation a -> Ordering
comparePermutation = compare `on` (functionRepr . to)

orderOfPermutation
  :: (Enum a, Bounded a) => Permutation -> Natural
orderOfPermutation Permutation{to = f} = go 1 fullSet
    where
      n = 1 + fromEnum (maxBound @a)
      fullSet = ISet.fromDistinctAscList [0 .. n - 1]
      
      f' :: Int -> Int
      f' = fromEnum . f . toEnum
      
      go :: Natural -> ISet.IntSet -> Natural
      go !ord elements = case ISet.minView elements of
        Nothing             -> ord
        Just (k, elements') ->
          let (period, elements'') = takeCycle k elements'
          in go (lcm period ord) elements''
      
      takeCycle :: Int -> ISet.IntSet -> (Natural, ISet.IntSet)
      takeCycle k = loop 1 (f' k)
        where
          loop !period j elements
            | j `ISet.member` elements      = loop (succ period) (f' j) (ISet.delete j elements)
            | {- j ∉ elements && -} j == k = (period, elements)
            | otherwise                     = error $ "Non-bijective: witness=toEnum " ++ show j
```

Reply:
I like that idea a little better. If people want those instances, they can newtype and write the instances themselves,